atc/exec/task_step.go

```diff
@@ -193,192 +205,65 @@ func (step *TaskStep) Run(ctx context.Context, state RunState) error {
-
-	var chosenWorker worker.Worker
-	var activeTasksLock lock.Lock
-	var elapsed time.Duration
-	outputWriter := step.delegate.Stdout()
-
-	for {
-		if step.strategy.ModifiesActiveTasks() {
-			var acquired bool
-			activeTasksLock, acquired, err = step.lockFactory.Acquire(logger, lock.NewTaskStepLockID())
-			if err != nil {
-				return err
-			}
-
-			if !acquired {
-				time.Sleep(time.Second)
-				continue
-			}
-		}
-
-		chosenWorker, err = step.workerPool.FindOrChooseWorkerForContainer(
-			ctx,
-			logger,
-			owner,
-			containerSpec,
-			workerSpec,
-			step.strategy,
-		)
-		if err != nil {
-			return err
-		}
+	processSpec := worker.TaskProcessSpec{
+		Path:         config.Run.Path,
+		Args:         config.Run.Args,
+		Dir:          config.Run.Dir,
+		StdoutWriter: step.delegate.Stdout(),
+		StderrWriter: step.delegate.Stderr(),
+	}
 
-		if step.strategy.ModifiesActiveTasks() {
-			waitWorker := time.Duration(5 * time.Second) // Workers polling frequency
-			if chosenWorker == nil {
-				logger.Info("no-worker-available")
-				err = activeTasksLock.Release()
-				if err != nil {
-					return err
-				}
-				if elapsed%time.Duration(time.Minute) == 0 { // Every minute report that it is still waiting
-					_, err := outputWriter.Write([]byte("All workers are busy at the moment, please stand-by.\n"))
-					if err != nil {
-						logger.Error("Failed to report status.", err)
-					}
-				}
-				elapsed += waitWorker
-				time.Sleep(waitWorker)
-				continue
-			}
-			err = chosenWorker.IncreaseActiveTasks()
-			if err != nil {
-				logger.Error("Failed to increase active tasks.", err)
-			}
-			logger.Info("increase-active-tasks")
-			err = activeTasksLock.Release()
-			if err != nil {
-				return err
-			}
-			if elapsed > 0 {
-				_, err := outputWriter.Write([]byte(fmt.Sprintf("Found a free worker after waiting %s.\n", elapsed)))
-				if err != nil {
-					logger.Error("Failed to report status.", err)
-				}
-			}
-			defer step.decreaseActiveTasks(logger, chosenWorker)
-		}
-		break
+	imageSpec := worker.ImageFetcherSpec{
+		ResourceTypes: resourceTypes,
+		Delegate:      step.delegate,
 	}
+	owner := db.NewBuildStepContainerOwner(step.metadata.BuildID, step.planID, step.metadata.TeamID)
 
-	container, err := chosenWorker.FindOrCreateContainer(
+	result := step.workerClient.RunTaskStep(
 		ctx,
 		logger,
-		step.delegate,
+		step.lockFactory,
 		owner,
-		step.containerMetadata,
 		containerSpec,
-		resourceTypes,
+		workerSpec,
+		step.strategy,
+		step.containerMetadata,
+		imageSpec,
+		processSpec,
+		events,
 	)
-	if err != nil {
-		return err
-	}
-
-	exitStatusProp, err := container.Property(taskExitStatusPropertyName)
-	if err == nil {
-		logger.Info("already-exited", lager.Data{"status": exitStatusProp})
-
-		status, err := strconv.Atoi(exitStatusProp)
-		if err != nil {
-			return err
-		}
 
-		step.succeeded = (status == 0)
+	close(events)
```

Comment:
Should it be `RunTaskStep`'s responsibility to close the channel since it is the producer?